src/preallocate.jl

```diff
@@ -0,0 +1,52 @@
+export preallocate
+
+struct PreallocatedMethod{F, Args <: Tuple, N, R}
+    f::F
+    ctx::NTuple{N, R}
+
+    function PreallocatedMethod{F, Args}(f::F, ctx::NTuple{N, R}) where {F, Args, N, R}
+        new{F, Args, N, R}(f, ctx)
+    end
+end
+
+function PreallocatedMethod(f::F, xs...) where F
+    x, record = record_allocations(f, xs...)
+    record = freeze(record)
+    records = (ntuple(_->copy(record), Threads.nthreads() - 1)..., record)
+    ctx = new_replay_ctx.(records)
+    return PreallocatedMethod{F, typeof(xs)}(f, ctx)
```

Comment:
Do you think its clearer if done all in one hit?
I think the -1 and splatting make it hard to read.

Also the splatting may cause problems on 32+ CPU machines (which I have people using)

```suggestion
    ctxs = ntuple(Threads.nthreads()) do ii
        new_replay_ctx(ii == 1 ? record : copy(record))
    end
    return PreallocatedMethod{F, typeof(xs)}(f, ctxs)
```